src/util/std_types.h

```diff
@@ -39,8 +31,7 @@ class bool_typet:public typet
   }
 };
 
-/*! \brief The NIL type
-*/
+/// The NIL type
```

Comment:
So ... what is it?

Reply:
Yes - in particular how is it different from the empty type! I think the answer is:

- nil type: an invalid type used where really `optional<typet>` should be used. 
- empty type `void` as in the subtype of a `void*` pointer.

Reply:
Though why is there a `void_typet` and an `empty_typet` I have no idea

Reply:
Yes, this is the explanation I got for `nil` vs `void/empty`. `void` and `empty` are two names for the same thing.

Reply:
I have created #2780 for `void_typet` vs `empty_typet`.